### PR TITLE
JAMES-3210 Handle Return-Path header in DSNBounce mailet

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

**Comment by chibenwa** [comment]
> If a DSN specialist can comment/suggest about it, it would be great.

I'm not one but I will give it a try...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 1]
```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
...
+    private List<MailAddress> getSenderAsList(Mail originalMail) throws MessagingException {
+        Optional<MailAddress> returnPath = Optional.ofNullable(originalMail.getMessage().getHeader(RETURN_PATH))
+                .flatMap(path -> path.length > 0 ? Optional.ofNullable(path[0]) : Optional.empty())
+                .map((Throwing.<String, MailAddress>function(MailAddress::new).sneakyThrow()));
 
-        if (getInitParameters().isDebug()) {
-            LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
+        if(returnPath.isPresent()) {
+            if (getInitParameters().isDebug()) {
+                LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", returnPath.get().asString());
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:340]: reverse path -> ReturnPath

Let's have a consistent vocabulary to not complexify domain knwoledge...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 2]
```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
...
-        MaybeSender reversePath = originalMail.getMaybeSender();
+    private List<MailAddress> getSenderAsList(Mail originalMail) throws MessagingException {
+        Optional<MailAddress> returnPath = Optional.ofNullable(originalMail.getMessage().getHeader(RETURN_PATH))
+                .flatMap(path -> path.length > 0 ? Optional.ofNullable(path[0]) : Optional.empty())
+                .map((Throwing.<String, MailAddress>function(MailAddress::new).sneakyThrow()));
 
-        if (getInitParameters().isDebug()) {
-            LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
+        if(returnPath.isPresent()) {
+            if (getInitParameters().isDebug()) {
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:339]: IMO we should deprecate these `getInitParameters().isDebug()` string everywhere in mailet configuration and expect people to directly configure there loggers. Can we open a JIRA issue about it?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 3]
```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
         }
     }
 
-    private List<MailAddress> getSenderAsList(Mail originalMail) {
-        MaybeSender reversePath = originalMail.getMaybeSender();
+    private List<MailAddress> getSenderAsList(Mail originalMail) throws MessagingException {
+        Optional<MailAddress> returnPath = Optional.ofNullable(originalMail.getMessage().getHeader(RETURN_PATH))
+                .flatMap(path -> path.length > 0 ? Optional.ofNullable(path[0]) : Optional.empty())
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:335]: Let me suggest `Arrays.stream(path).findFirst()` that allows getting rid of the condition?

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:335]: I think we are far from other as mail address header parsing is complex...

Do we handle the following cases?

```
Return-Path: inv@lid@domain.tld
```

(I expect a fallback on the sender)

```
Return-Path: <wrapped@domain.tld>
```

(Eg mails from JIRA follows this pattern)

```
Return-Path: "Wrapped" <wrapped@domain.tld>
```

```
Return-Path: "Wrapped" 
    <wrapped@domain.tld>
```

& also encoding handling...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 4]
```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
...
-            LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
+        if(returnPath.isPresent()) {
+            if (getInitParameters().isDebug()) {
+                LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", returnPath.get().asString());
+            }
+            return ImmutableList.of(returnPath.get());
+        } else {
+            MaybeSender reversePath = originalMail.getMaybeSender();
+            if (getInitParameters().isDebug()) {
+                LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:346]: You don't have a returnPath and you default to the sender.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 5]
```diff
@@ -321,13 +330,23 @@ private String getRemoteAddr() {
...
 
-        if (getInitParameters().isDebug()) {
-            LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", reversePath.asString());
+        if(returnPath.isPresent()) {
+            if (getInitParameters().isDebug()) {
+                LOGGER.debug("Processing a bounce request for a message with a reverse path.  The bounce will be sent to {}", returnPath.get().asString());
+            }
+            return ImmutableList.of(returnPath.get());
+        } else {
+            MaybeSender reversePath = originalMail.getMaybeSender();
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:344]: Maybe we could falling back to To MIME fields before falling back to the SMTP envelope sender?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java [hunk 6]
```diff
@@ -253,9 +259,11 @@ public void service(Mail originalMail) throws MessagingException {
     private void trySendBounce(Mail originalMail) throws MessagingException {
         MailImpl newMail = MailImpl.duplicate(originalMail);
         try {
+            List<MailAddress> bounceReceiver = getSenderAsList(originalMail);
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:262]: You don't get things, you somehow "compute" things.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java (hunk 5 above)
Comment (rouazana) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:344]: are you sure it makes sense for a DSN? (I really don't know, but would be happy to read some clue about this)

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:344]: (no idea: I'm not a DSN expert)

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:344]: Actually https://tools.ietf.org/html/rfc3461 states:

```
   The DSN MUST be addressed (in both the message header and the
   transport envelope) to the return address from the transport envelope
   which accompanied the original message for which the DSN was
   generated.  (For a message that arrived via SMTP, the envelope return
   address appears in the MAIL FROM command.)
```

:-/

**Comment by Arsnael** [comment]
```
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [INFO] There are 5 errors reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [ERROR] src/main/java/org/apache/james/transport/mailets/DSNBounce.java:[33,8] (imports) UnusedImports: Unused import - java.util.stream.Stream.
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [ERROR] src/main/java/org/apache/james/transport/mailets/DSNBounce.java:[42,8] (imports) UnusedImports: Unused import - org.apache.commons.lang3.ArrayUtils.
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [ERROR] src/main/java/org/apache/james/transport/mailets/DSNBounce.java:[63,8] (imports) UnusedImports: Unused import - org.apache.james.util.OptionalUtils.
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [ERROR] src/main/java/org/apache/james/transport/mailets/DSNBounce.java:[74,8] (imports) UnusedImports: Unused import - com.github.fge.lambdas.functions.ThrowingFunction.
[28959c305289df3ecce9ed31026f9f30c6ce6e06] [ERROR] src/main/java/org/apache/james/transport/mailets/DSNBounce.java:[338,9] (whitespace) WhitespaceAround: WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)
```

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/BounceIntegrationTest.java [hunk 7]
```diff
@@ -71,177 +76,146 @@ public void tearDown() {
...
+        messageSender.connect(LOCALHOST_IP, jamesServer.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessage(SENDER, RECIPIENT);
+
+        testIMAPClient.connect(LOCALHOST_IP, jamesServer.getProbe(ImapGuiceProbe.class).getImapPort())
+            .login(SENDER, SENDER_PASSWORD)
+            .select(TestIMAPClient.INBOX)
+            .awaitMessage(awaitAtMostOneMinute);
+    }
+
+    private void setup(Class<? extends Mailet> mailet, Pair<String, String>... additionalProperties) throws Exception {
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/BounceIntegrationTest.java:90]: Please rely on a strong type here. `MailetConfiguration` (or it's builder) should do the work.

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/BounceIntegrationTest.java [hunk 8]
```diff
@@ -71,177 +76,146 @@ public void tearDown() {
...
+                .addUser(POSTMASTER, POSTMASTER_PASSWORD);
+    }
+
+    private void setupForBounce() throws Exception {
+        setup(Bounce.class);
+    }
+
+    private void setupForDsnBounce() throws Exception {
+        setup(DSNBounce.class);
+    }
```
Comment (chibenwa) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/BounceIntegrationTest.java:123]: IMO these method are of little use...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java (hunk 5 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:340]: The quote is taken from https://tools.ietf.org/html/rfc3464 `An Extensible Message Format for Delivery Status Notifications` https://tools.ietf.org/html/rfc3464#section-2

(Sorry I messed up)

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DSNBounce.java:340]: https://tools.ietf.org/html/rfc3834 `Recommendations for Automatic Responses to Electronic Mail`

```
In general, automatic responses SHOULD be sent to the Return-Path
   field if generated after delivery.  If the response is generated
   prior to delivery, the response SHOULD be sent to the reverse-path
   from the SMTP MAIL FROM command, or (in a non-SMTP system) to the
   envelope return address which serves as the destination for non-
   delivery reports.
```

**Comment by remk** [comment]
It seems that this PR should be closed as the return-path should not be used in DSN usage.
I will extract the tests refactoring in a new one.

REGARDING RFC 3834 "Recommendations for Automatic Responses to Electronic Mail" which specify the usage of the return-path field  was quoted :

3.2.1.  Use of DSNs and MDNs instead of this specification

>    In general, it is appropriate to use Delivery Status Notifications
>    (DSNs) for responses that are generated by the mail transport system
>    as a result of attempts to relay, forward, or deliver mail, and only
>    when the purpose of that response is to provide the sender of the
>    subject message with information about the status of that mail
>    delivery.  For instance, a "virus scanner" which is activated by a
>    mail delivery process to filter harmful content prior to delivery,
>    could return a DSN with the Action field set to "failed" with a
>    Status code of 5.7.1 (Delivery not authorized, message refused) if
>    the entire message was not delivered due to security reasons; or it
>    could return a DSN with the Action field set to "relayed" or
>    "delivered" (as appropriate) with a Status code set to 2.6.4
>    (conversion with loss performed) if the message was relayed or
>    delivered with the presumably harmful content removed.  The DSN
>    specification [I2.RFC3464], rather than this document, governs the
>    generation and format of DSNs.

And the RFC 3464 specifying the DSN : 

>    The DSN MUST be addressed (in both the message header and the
>    transport envelope) to the return address from the transport envelope
>    which accompanied the original message for which the DSN was
>    generated.  (For a message that arrived via SMTP, the envelope return
>    address appears in the MAIL FROM command.)

**Comment by rouazana** [comment]
Sorry, I don't get why it is closed. The second quote from the RFC tells that the return address should be the one from the envelope. It's currently not the case, is it?

**Pull request closed without merging**